lib/spack/spack/util/url.py

```diff
@@ -343,7 +343,9 @@ def parse_git_url(url):
 
 
 def require_url_format(url):
-    ut = re.search(r'^(file://|http://|https://|ftp://|s3://|gs://|ssh://|git://|/)', url)
+    # `:pserver:` is the prefix used for remote CVS repositories.
+    # The CVS protocol predates the URL standard, hence the non-standard shape.
+    ut = re.search(r'^(file://|http://|https://|ftp://|s3://|gs://|ssh://|git://|:pserver:|/)', url)
```

Comment:
I'm guessing this had to be done because functions like `CvsFetchStrategy.mirror_id` would fail (since they call `url.parse`) - as long as CVS references otherwise behave like URLs, then this makes sense. Otherwise, we might want to do special-case parsing for them. 

Another reason to do special-case parsing is to support other methods.

Is `CvsFetchStrategy.mirror_id` the only function that would need to be updated to avoid the errors you were seeing?

Reply:
A "typical" CVS repository location would be `:pserver:anonymous:anonymous@octane.obspm.fr:/cvsroot%module=Lorene`. There is the `:pserver:` prefix, a user name, a password, a host name, a path, and a parameter. (The parameter is handled by Spack, as cvs expects the repository location and module to be passed as two separate parameters.) It's quite similar to a url.

Reply:
> It's quite similar to a url

Since this doesn't match the specification from https://datatracker.ietf.org/doc/html/rfc3986#section-3, I was concerned that processing the repository location with `urllib` functions may have undefined behavior AFAIK. That being said, some of this is due to technical debt in the `FetchStrategy` "interface" (i.e. the `url_attr` for `CvsFetchStrategy`), and this PR doesn't make anything worse since all these functions were called before; in fact if handling my other concerns is overly-time-consuming, this is ideal since it highlights a conceptual mismatch.

I think the bigger possible concern is that `require_url_format` would need additional updates to "support" other CVS retrieval methods (i.e. other than `:pserver:`), so I was curious if the following was true:

> Is CvsFetchStrategy.mirror_id the only function that would need to be updated to avoid the errors you were seeing?

If avoiding these indirect calls to `require_url_format` is difficult, then I think the way you did it here is the best way to do it for now (since there doesn't seem to be a need for repository locations other than `:pserver:`). But if this is as simple as just editing `mirror_id` then I think it would be ideal to do that here

If the code stays as is, I think a TODO ought to be added not to treat `CvsFetchStrategy.url_attr` as a URL (and to call it something different).

Reply:
It is fine to return `None` in the sense that the mirroring logic should accept that (in that case it won't generate repository-unique identifying names), although I was thinking a special-case regex (that knows how to parse the path from a CVS repository location) would be ideal. Based on the definition in the section "2.9 Remote repositories" at https://www.gnu.org/software/trans-coord/manual/cvs/cvs.html, everything after the last `:` forms the path (minus a possible port-number at the beginning):

```
# Special-case handling because this is not actually a URL
elements = self.url.split(':')
final = elements[-1]
path_re = r'^\d*([^0-9]+$)'
path = re.match(path_re, final).groups(1)
```

Reply:
Done.